ci/travis/bazel.py

```diff
@@ -0,0 +1,227 @@
+            lambda m: textproto_format(*(m.groups() + (json_encoder, ))),
+            next_line)
+        outputs.append(prev_tail + next_line)
+        if line == b"}":
+            yield b"".join(outputs)
+            del outputs[:]
+        prev_comma = line != b"}" and (next_line.endswith(b"]")
+                                       or next_line.endswith(b"\""))
+        prev_tail = tail
+    if len(outputs) > 0:
+        yield b"".join(outputs)
+        del outputs[:]
+
+
+def textproto_parse(stream, encoding, json_encoder):
+    for item in textproto_split(stream, json_encoder):
+        yield json.loads(item)
+
+
+class Bazel(object):
+    encoding = "utf-8"
+
+    def __init__(self, program=None):
+        if program is None:
+            program = os.getenv("BAZEL_EXECUTABLE", "bazel")
+        self.argv = (program, )
+        self.extra_args = ("--show_progress=no", )
+
+    def _call(self, command, *args):
+        return subprocess.check_output(
+            self.argv + (command, ) + args[:1] + self.extra_args + args[1:],
+            stdin=subprocess.PIPE)
+
+    def info(self, *args):
+        result = OrderedDict()
+        for line in self._call("info", *args).splitlines():
+            (key, value) = line.split(b":", 1)
+            if value.startswith(b" "):
+                value = value[1:]
+            result[key.decode(self.encoding)] = value.decode(self.encoding)
+        return result
+
+    def aquery(self, *args):
+        lines = self._call("aquery", "--output=textproto", *args).splitlines()
+        return textproto_parse(lines, self.encoding, json.JSONEncoder())
+
+
+def parse_aquery_shell_calls(aquery_results):
+    for (key, val) in aquery_results:
+        if key == "actions":
+            [mnemonic] = [pair[1] for pair in val if pair[0] == "mnemonic"]
+            if mnemonic == "Genrule":
+                yield [pair[1] for pair in val if pair[0] == "arguments"]
+
+
+def parse_aquery_output_artifacts(aquery_results):
+    artifacts = {}
+    for (key, val) in aquery_results:
+        if key == "artifacts":
+            [artifact_id] = [pair[1] for pair in val if pair[0] == "id"]
+            [exec_path] = [pair[1] for pair in val if pair[0] == "exec_path"]
+            artifacts[artifact_id] = exec_path
+        elif key == "actions":
+            output_ids = [pair[1] for pair in val if pair[0] == "output_ids"]
+            for output_id in output_ids:
+                yield artifacts[output_id]
+
+
+def textproto2json(infile, outfile):
+    json_encoder = json.JSONEncoder(indent=2)
+    encoding = "utf-8"
+    for obj in textproto_parse(infile, encoding, json_encoder):
+        outfile.write((json_encoder.encode(obj) + "\n").encode(encoding))
+
+
+def preclean(bazel_aquery):
+    """Cleans up any genrule() outputs for the provided target(s).
+
+    This is useful for forcing genrule actions to re-run, because the _true_
+    outputs of those actions can include a larger set of files (e.g. files
```

Comment:
If this is the case, the genrule is probably broken. Calling out where this happens in the BUILD.bazel would be helpful (so that it may at least be a caution, if not fixed, and preclean obviated)

Reply:
The genrules are indeed "broken", but that has always been the case, and is unrelated to this PR. The reason is that we want to put the generated files into the worktree and Bazel inherently doesn't support this. There's no way to avoid the precleaning short of revamping the entire build flow and avoiding modifying the worktree (which I'm not against if it's possible, but is orthogonal to this PR).

Reply:
This isn't new logic btw. It's equivalent to what was already in the other shell script I deleted. I'm just moving it from Bash to Python to deduplicate the textproto parsing logic since I had to solve that problem in here in Python too.